include/kllvm/util/temporary_file.h

```diff
@@ -0,0 +1,50 @@
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <unistd.h>
+
+typedef std::unique_ptr<FILE, decltype(&fclose)> temp_c_file_type;
+
+class temporary_file {
+private:
+  int temp_fd;
+  std::string temp_filename;
+  temp_c_file_type temp_c_file;
+  std::optional<std::ofstream> temp_cpp_file;
+
+public:
+  temporary_file(std::string template_name)
+      : temp_c_file(nullptr, &fclose) {
+    temp_fd = mkstemp(template_name.data());
+    temp_filename = template_name;
+
+    if (temp_fd == -1) {
+      std::runtime_error("Could not create temporary file!");
+    }
+  }
```

Comment:
There's a slightly nuanced improvement we can make to this constructor. As written, we're default-initializing `temp_filename`, then assigning it a new value with its `operator=(std::string const&)`. What we can do instead is member-initialize it, then call `mkstemp` on the member variable rather than the constructor argument. The code will look something like:
```c++
  temporary_file(std::string template_name)
      : temp_filename(template_name)
      , temp_c_file(nullptr, &fclose)
      , temp_fd(mkstemp(temp_filename.data())) {
    if (temp_fd == -1) {
      std::runtime_error("Could not create temporary file!");
    }
  }
```

Note that you'll need to reorder the class members to get this to compile.